### ISSUE TITLE
Using tuning_run to experiment with neural network architecture

### ISSUE THREAD

**Issue by ericricky** [rstudio/tfruns#32]
I was experimenting with using `tuning_run` to collect data on the performance of models with different numbers of hidden layers, and I wanted to ask about some of the issues that came up. Currently, tfrun requires flags of length one and has stringent requirements on the form of the hyperparameter subspace to search - i.e. it can only search (possibly random subsets of) the space of all combinations of values specified in individual vectors for each flag. This makes it difficult to use `tuning_run` to experiment with different numbers of hidden layers and different numbers of units in each hidden layer. First, is there any downside to using `tuning_run` in this way? Is it more commonly used to optimize simpler parameters?

The most natural way I can think of to control the number of hidden layers and hidden units through flags is to have a single single flag which is a vector of integers which gives the number of units for each layer. Since flags need to be length one, another approach is to use an integer flag `num_hidden_layers` for the number of hidden layers, and flags `hidden_units_1`, `hidden_units_2`, etc for the number of hidden units in each layer. The problem then is that you don't want to execute multiple training runs where `num_hidden_layers` is constant and `hidden_units_x` changes but isn't used. You can probably get around this by using an `if` statement in your model script to not fit the model in these cases. Or you can get around the problem altogether by parameterizing the number of hidden layers and number of units in each layer by a single integer flag, which isn't hard but is unnatural and hard to read.

Anyways, I think this raises some more general issues. Sometimes, as in the case of `hidden_units_x` and `num_hidden_layers` above, flags are naturally nested in other flags. As another example, someone might use a learning rate flag and an optimizer flag and only want to try certain values of learning rate for a certain optimizer. Any thoughts on this? These issues could probably be solved by introducing a way to specify your own set of parameter combinations to try, as in https://github.com/rstudio/tfruns/issues/30. Maybe the user could specify a data frame of flags, with one column per flag. Would you be interested in a pull request with a feature like this?
